### ISSUE TITLE
sql: table scans from equality matches on unique columns should use kv Get, not Scan

### ISSUE THREAD

**Issue by knz** [cockroachdb/cockroach#37269]
Currently, table lookups use Scan requests on a row prefix to get data from kv, even when the logical plan suggests that just 1 row is to be looked up at a specfic key prefix (common case: equality match on a unique column). This is perfectly reasonable, but it leaves performance on the table in most OLTP workloads.

When possible, 1-row kv lookups should instead use Get requests - one Get request per column family in the table (and thus 1 get request overall in the common case where there is just 1 column family).

Note: This is different from the current "limit 1" annotation which indicates that although the span may encompass multiple PK values, the query expects just 1 row to be present in the range. In that case, a Scan must be used because there is no certainty that the value to get is present at the start key in the span.

The issue here is for the specific case where the query planner knows that the start key in the span is also the key at which the row will be found. Currently this cannot be expressed in logical/physical plans.

To achieve this two components should be modified:
- the logical plan node for table scans must receive a new field to indicate that the row is expected at the start key of the span
- the query optimizer must learn to populate this field
- the table reader processor in distsql must learn to use it to issue kv Get requests instead of Scan 

Solving this should have a noticeable impact on TPCC performance.

**Comment by jordanlewis** [comment]
@nvanbenschoten we talked about this and from what I recall came to the conclusion that a Get won't lead to any actual performance benefits over the equivalent tight Scan request. Can you confirm?

**Comment by nvanbenschoten** [comment]
I'm not completely sure. For ScanRequests, we initialize the RocksDB iterator with the scan's upper and lower bound, so it's likely possible that we set the prefix bloom filter up so that it can be used for ScanRequests over a single row, but I doubt that actually works at the moment. At a minimum, we'll probably need to bake some understanding of column families into the bloom prefix for it to work correctly in this situation.

@petermattis we've discussed this in the past and I've been surprised that you didn't think this was particularly important. Are we missing something here?

**Comment by petermattis** [comment]
RocksDB only uses bloom filters when prefix iteration is set. For CockroachDB, this is only done for `MVCCGet`, never for `MVCCScan`. I believe @ajkr said there was thinking going on about how to use prefix bloom filters for general scans if the lower and upper bound were set appropriately.

> At a minimum, we'll probably need to bake some understanding of column families into the bloom prefix for it to work correctly in this situation.

Are you imagining we'd generate the bloom filter on the "row prefix"? 

> @petermattis we've discussed this in the past and I've been surprised that you didn't think this was particularly important. Are we missing something here?

So many of our workloads are in cache, which lessens the importance of the bloom filter. Happy to chat more about this in person.

**Comment by nvanbenschoten** [comment]
> Are you imagining we'd generate the bloom filter on the "row prefix"?

Yes, that's what I was picturing. To do so, we'd need to be careful about interleaved tables, but I _think_ that would be manageable.

> So many of our workloads are in cache, which lessens the importance of the bloom filter. Happy to chat more about this in person.

I'm surprised that we would be ok with passing up on optimizations for the cases where this isn't true. If we truly believed that all of the datasets that we care about will fit in memory then we picked the wrong storage architecture.

I'd like to talk about this more in person. Specifically, I'm curious whether you think this is ever worth exploring and whether there's an intern project sitting here.

**Comment by petermattis** [comment]
> I'm surprised that we would be ok with passing up on optimizations for the cases where this isn't true. If we truly believed that all of the datasets that we care about will fit in memory then we picked the wrong storage architecture.

Let me rephrase my position. There very well might be significant improvements that can be had when the working set does not fit in cache. I think we should see how often those workloads are occurring and come up with a representative workload. And then we should analyze it. Essentially, I want to put the workload first, not the performance change.

> Yes, that's what I was picturing. To do so, we'd need to be careful about interleaved tables, but I think that would be manageable.

Would we maintain two bloom filters: one on the user key and one on the "row prefix"? Or only the one on the "row prefix"? The former seems expensive from a memory perspective, the could have an impact on operations which rely on the current bloom filter, though I'm not sure how beneficial it is. I'd be interested in knowing how performance is affected if we disable the current bloom filter. That would be an easy test, and give some sense of the impact of the change proposed here.

**Comment by ajkr** [comment]
> Would we maintain two bloom filters: one on the user key and one on the "row prefix"? Or only the one on the "row prefix"?

I'm confused about whether we know the key for the first row in the range or we only know the row prefix. Until this comment I was thinking the former. Assuming that's the case, I agree with using `MVCCGet()` here. If we set `iterate_lower_bound` and `iterate_upper_bound` properly the iterator can use the prefix bloom filters we already have (and almost never use according to the statistics!).

**Comment by petermattis** [comment]
> If we set iterate_lower_bound and iterate_upper_bound properly the iterator can use the prefix bloom filters we already have (and almost never use according to the statistics!).

I just tracked down how this works (`FullFilterBlockReader::RangeMayExist`) and it appears to only use the `iterate_upper_bound`. Is that correct? I need to understand how this code a bit better, but it might obviate the need to try to use `MVCCGet` as long as we provided a tight `iterate_upper_bound` (which we should be able to do).

**Comment by ajkr** [comment]
Yes that looks correct. I assumed it required both lower and upper bounds but it looks like only the latter is necessary. Does this mean there's an assumption that an iterator with upper bound set won't move backwards? Confusing.

**Comment by petermattis** [comment]
> Yes that looks correct. I assumed it required both lower and upper bounds but it looks like only the latter is necessary. Does this mean there's an assumption that an iterator with upper bound set won't move backwards? Confusing.

Huh, `SeekForPrev` uses the same logic. I wonder if that is a bug.

**Comment by knz** [comment]
still relevant

**Comment by jordanlewis** [comment]
@knz actually I closed this in #61583 😎
